[Elixir] Implementing a Concept Exercise Guide (#888)
exercism/v3: languages/elixir/reference/implementing-a-concept-exercise.md

# How to implement an Elixir Concept Exercise

This document describes how to implement a Concept Exercise for the Elixir track. As this document is generic, the following placeholders are used:

- `$slug`: the name of the exercise in snake_case (e.g. `anonymous-methods`).

> - `$concepts`: the Concepts the exercise is about (e.g. `loops`),

Author: ```suggestion
- `$elixir_slug`: `$slug` converted to snake-case from kebab-case,
- `$concepts`: the Concepts the exercise is about (e.g. `loops`),
```

- `$concept-1`: a single Concept slug,
- `$prerequisite-n`: a single Concept slug,
- `$uuid`: a _new_ v4 UUID (random! and unique to all other concept and practice exercises)
- `$first-and-last-name`: your first and last name (e.g. `Tim Austin`)
- `$git-email`: the email address you use for git (e.g. `tim@neenjaw.com`)

Before implementing the exercise, please make sure you have a good understanding of what the exercise should be teaching (and what not). This information can be found in the exercise's GitHub issue. Having done this, please read the [Elixir Concept exercises introduction][concept-exercises]. If you have come up with something completely new, create a new issue _first_ so we can discuss the Concept Exercise.

To implement a Concept Exercise, the following files must be created:

```text
languages
└── elixir
    └── exercises
        └── concept
            └── $slug
                ├── .docs
                │   ├── instructions.md
                │   ├── introduction.md
                │   ├── hints.md
                │   └── after.md (optional)
                ├── .meta
                │   ├── config.json
                │   └── design.md
                ├── lib

>                 │   ├── example.js

Reviewer: ```suggestion
                │   ├── example.ex
```

                │   └── $slug.ex
                ├── mix.exs
                ├── mix.lock
                └── test
                    ├── $slug_test.exs

>                     └── test_test.exs

Reviewer: Are there multiple test files?

Author: ```suggestion
                    └── test_helper.exs
```

Author: I made a spelling error, the test helper is generated by the mix project generation.

The hints should not spell out the solution, but instead point to a resource describing the solution (e.g. linking to documentation for the function to use).

## Step 4: add .docs/after.md (optional)

Once the student completes the exercise they will be shown this file, which should provide them with a summary of what the exercise aimed to teach. This document can also link to any additional resources that might be interesting to the student in the context of the exercise.

## Step 5: add .meta/config.json

This file contains metadata for the exercise. The metadata includes paths to the test file and the solution files, as well as a list of all the tests and the method they are calling.

## Step 6: update languages/Elixir/config.json

An entry should be added to the track's `config.json` file for the new Concept Exercise:

```json
{
  ...
  "exercises": {
    "concept": [
      ...
      {
        "slug": "$slug",
        "uuid": "$uuid",
        "concepts": ["$concept-1"],
        "prerequisites": ["$prerequisite-1", "$prerequisite-2"]
      }
    ]
  }
}
```

## Step 7: adding track-specific files

> The configuration files may be copied from another exercise. But the structure of an exercise is a generic elixir mix-project.

Reviewer: Maybe I'm reading this wrong, but can't the files be copied _because_ it is generic elixir mix project?

```suggestion
The configuration files may be copied from another exercise, as the structure of an exercise is a generic elixir mix-project.
```

Author: They can be, but it's probably better to make a new one with `mix new $slug_with_underscores_not_dashes` then move the contents to the `$slug` directory

Author: ```suggestion
The configuration files may be copied from another exercise. But it would be recommended to use `mix new $slug` where `$slug` uses underscores `_` instead of dashes `-` for file generation, the move the generated structure to the `$slug` directory to match the structure in this guide.
```

Reviewer: Okay good!

Now create the following three files:

> - `$slug.ex`. the stub implementation file, which is the starting point for students to work on the exercise.

Reviewer: ```suggestion
- `lib/$slug.ex`. the stub implementation file, which is the starting point for students to work on the exercise.
```

> - `$slug_test.ex`: the test suite.

Reviewer: ```suggestion
- `test/$slug_test.ex`: the test suite.
```

> - `example.ex`: an example implementation that passes all the tests.

Reviewer: ```suggestion
- `lib/example.ex`: an example implementation that passes all the tests.
```
